### PR TITLE
fix: improvements in token balance fetching methods

### PR DESCRIPTION
**Pull request by Abhikumar98**
# FEATURE

Improve SDK getTokenBalancesForChains and getTokenBalances methods [LF-3225]

## TL;DR

Add a boolean parameter to the token balance methods to allow users to hide zero balance tokens. Also, improve error handling for provider errors during token balance fetching.

## DESCRIPTION

This feature improves the SDK's `getTokenBalancesForChains` and `getTokenBalances` methods by adding a boolean parameter. This parameter allows users to specify whether they want to hide zero balance tokens or not. By default, all tokens are returned, but users can choose to exclude tokens with zero balances by setting the boolean parameter to `true`.

Additionally, this feature enhances the error handling for provider errors during token balance fetching. It throws the correct validation error when encountering issues with the provider during the retrieval of token balances.

## STEPS TO TEST

1. Call the `getTokenBalancesForChains` and `getTokenBalances` methods with the boolean parameter set to `true` to hide zero balance tokens.
2. Verify that the returned token balances exclude tokens with zero balances.

## EXPECTED RESULTS

- When setting the boolean parameter to `true`, the token balance methods should return balances excluding zero balance tokens.
- When encountering provider errors during token balance fetching, the methods should throw the correct validation error.

## CHECKLIST

- [ ] I have performed a self-review of my code.
- [ ] I have commented the code I wrote to ensure clarity.
- [ ] I have added tests that cover the new functionality.
- [ ] I have made sure that existing tests are still passing.

[LF-3225]: https://lifi.atlassian.net/browse/LF-3225?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by chybisov** [comment]
Can we just hide zero balances without any additional flag? The question here https://lifi.atlassian.net/browse/LF-3225 was mainly for the widget and if it is possible to hide them completely. 🙂

**Comment by Abhikumar98** [comment]
> Can we just hide zero balances without any additional flag? The question here https://lifi.atlassian.net/browse/LF-3225 was mainly for the widget and if it is possible to hide them completely. 🙂

Yes, we can filter out the zero balance tokens by default, just to be sure, there is no other use case where we'd have to show it?

File: src/balance/utils.ts [hunk 1]
```diff
@@ -158,8 +160,11 @@ const getBalancesFromProvider = async (
           .toString()
         blockNumber = balance.blockNumber
       } catch (e) {
-        // eslint-disable-next-line no-console
         console.warn(e)
+        throw new ProviderError(
+          (e as any).code,
+          getMessageFromCode((e as any).code)
+        )
```
[the remark this answers is not in the document]
Comment (Abhikumar98) [reply src/balance/utils.ts:167]: I don't think this method isn't being used in the widget, we are directly fetching the token details from backend.

Lmk if I'm missing something here 🙂

**Comment by chybisov** [comment]
> > Can we just hide zero balances without any additional flag? The question here https://lifi.atlassian.net/browse/LF-3225 was mainly for the widget and if it is possible to hide them completely. 🙂
> 
> Yes, we can filter out the zero balance tokens by default, just to be sure, there is no other use case where we'd have to show it?

It's better to double-check where we use this.

File: src/balance/utils.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (chybisov) [reply src/balance/utils.ts:167]: We fetch balances via SDK, so it can be used in the widget.

**Comment by chybisov** [comment]
This will be integrated into the new version after migration.
